### PR TITLE
[MRG+1] Showcase example: (kind of mandatory) Mandelbrot set

### PR DESCRIPTION
**Pull request by rougier**
This is a rendering of the Mandelbrot set. I think it is reasonably good looking (using normalized recount, adapted normalized colormap and shading).

mandelbrot

### CODE REVIEW

**Comment by stonebig** [comment]
I get these error message testing under python-3.6.0b3, don't know if it's important:

C:/WinPython/basedir36/build/winpython-64bit-3.6.x.0/notebooks/mandelbrot.py:41: RuntimeWarning: invalid value encountered in log
  M = np.nan_to_num(N + 1 - np.log(np.log(abs(Z)))/np.log(2) + log_horizon)
C:\WinPython\basedir36\build\winpython-64bit-3.6.x.0\python-3.6.0b3.amd64\lib\site-packages\matplotlib\font_manager.py:1288: UserWarning: findfont: Font family ['Source Sans Pro Light'] not found. Falling back to Bitstream Vera Sans
  (prop.get_family(), self.defaultFamily[fontext]))

**Comment by rougier** [comment]
@stonebig I removed the second warning (font specification).

File: examples/showcase/mandelbrot.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    M = light.shade(M, cmap=plt.cm.hot, vert_exag=1.5,
+                    norm=colors.PowerNorm(0.3), blend_mode='hsv')
+    plt.imshow(M, extent=[xmin, xmax, ymin, ymax], interpolation="bicubic")
+    ax.set_xticks([])
+    ax.set_yticks([])
+
+    # Some advertisement for matplotlib
+    ax.text(xmin+0.025, ymin+0.025,
+            "The Mandelbrot fractal set\n"
+            "Rendered with matplotlib 2.0, 2016 — http://www.matplotlib.org",
```
Comment (anntzer) [inline examples/showcase/mandelbrot.py]: Dynamically fetch the version and the year here, perhaps?

Comment (rougier) [inline examples/showcase/mandelbrot.py]: Yes, good point.

**Comment by NelleV** [comment]
Hi Nicolas,

Can you also please add a docstring with a title and a description of the example? It should be formatted as follow:
```python
"""
===============================
Colormaps alter your perception
===============================

Here I plot the function

.. math:: f(x, y) = \sin(x) + \cos(y)

with different colormaps. Look at how colormaps alter your perception!
"""
```

**Comment by rougier** [comment]
Hi Nelle,

Sure.

File: examples/showcase/mandelbrot.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    Z = np.zeros(C.shape, np.complex64)
+    for n in range(maxiter):
+        I = np.less(abs(Z), horizon)
+        N[I] = n
+        Z[I] = Z[I]**2 + C[I]
+    N[N == maxiter-1] = 0
+    return Z, N
+
+
+if __name__ == '__main__':
```
Comment (QuLogic) [inline examples/showcase/mandelbrot.py]: I don't think this conditional block is necessary.

File: examples/showcase/mandelbrot.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    xmin, xmax, xn = -2.25, +0.75, 3000/2
+    ymin, ymax, yn = -1.25, +1.25, 2500/2
+    maxiter = 200
+    horizon = 2.0 ** 40
+    log_horizon = np.log(np.log(horizon))/np.log(2)
+    Z, N = mandelbrot_set(xmin, xmax, ymin, ymax, xn, yn, maxiter, horizon)
+
+    # Normalized recount as explained in:
+    # http://linas.org/art-gallery/escape/smooth.html
```
Comment (QuLogic) [inline examples/showcase/mandelbrot.py]: https

File: examples/showcase/mandelbrot.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    # https://www.ibm.com/developerworks/community/blogs/jfp/entry/My_Christmas_Gift
+
+    # This line will generate warnings for null values but it is faster to
+    # process them afterwards using the nan_to_num
+    M = np.nan_to_num(N + 1 - np.log(np.log(abs(Z)))/np.log(2) + log_horizon)
```
Comment (QuLogic) [inline examples/showcase/mandelbrot.py]:
```python
with np.errstate(invalid='ignore'):
    M = np.nan_to_num(...)
```
?

File: examples/showcase/mandelbrot.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    dpi = 72
```
Comment (QuLogic) [inline examples/showcase/mandelbrot.py]: Is changing dpi important?

File: examples/showcase/mandelbrot.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    width = 10
+    height = 10*yn/xn
+    fig = plt.figure(figsize=(width, height), dpi=dpi)
+    ax = fig.add_axes([0.0, 0.0, 1.0, 1.0], frameon=False, aspect=1)
+
+    # Shaded rendering
+    light = colors.LightSource(azdeg=315, altdeg=10)
+    M = light.shade(M, cmap=plt.cm.hot, vert_exag=1.5,
```
Comment (QuLogic) [inline examples/showcase/mandelbrot.py]: Should we use our fancy new colour maps (magma, inferno, or plasma)? I'm not sure how well they work with shading though.

File: examples/showcase/mandelbrot.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                    norm=colors.PowerNorm(0.3), blend_mode='hsv')
+    plt.imshow(M, extent=[xmin, xmax, ymin, ymax], interpolation="bicubic")
+    ax.set_xticks([])
+    ax.set_yticks([])
+
+    # Some advertisement for matplotlib
+    year = time.strftime("%Y")
+    major, minor, micro = matplotlib.__version__.split('.')
```
Comment (QuLogic) [inline examples/showcase/mandelbrot.py]: You never know what might be in the local version identifier, so you should add `maxsplit=3` here.

File: examples/showcase/mandelbrot.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    text = ("The Mandelbrot fractal set\n"
+            "Rendered with matplotlib %s.%s, %s — http://matplotlib.org"
+            % (major, minor, year))
+    ax.text(xmin+.025, ymin+.025, text, color="white", fontsize=12, alpha=0.5)
+
+    # plt.savefig("mandelbrot.png")
```
Comment (QuLogic) [inline examples/showcase/mandelbrot.py]: Remove.

File: examples/showcase/mandelbrot.py (hunk 4 above)
Comment (rougier) [inline examples/showcase/mandelbrot.py]: Yes.

File: examples/showcase/mandelbrot.py (hunk 6 above)
Comment (rougier) [inline examples/showcase/mandelbrot.py]: I tried but output was not as nice as with the hot colormap.

File: examples/showcase/mandelbrot.py (hunk 8 above)
Comment (rougier) [inline examples/showcase/mandelbrot.py]: Will do.

File: examples/showcase/mandelbrot.py (hunk 2 above)
Comment (rougier) [inline examples/showcase/mandelbrot.py]: It doesn't hurt and allow to import mandelbrot if you want to experience with it.

File: examples/showcase/mandelbrot.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    light = colors.LightSource(azdeg=315, altdeg=10)
+    M = light.shade(M, cmap=plt.cm.hot, vert_exag=1.5,
+                    norm=colors.PowerNorm(0.3), blend_mode='hsv')
+    plt.imshow(M, extent=[xmin, xmax, ymin, ymax], interpolation="bicubic")
+    ax.set_xticks([])
+    ax.set_yticks([])
+
+    # Some advertisement for matplotlib
+    year = time.strftime("%Y")
+    major, minor, micro = matplotlib.__version__.split('.', maxsplit=3)
```
Comment (QuLogic) [inline examples/showcase/mandelbrot.py]: Sorry, got the number wrong; should be 2.

**Comment by NelleV** [comment]
Thanks @rougier ! This looks goot to me!

File: examples/showcase/mandelbrot.py (hunk 5 above)
Comment (QuLogic) [inline examples/showcase/mandelbrot.py]: I'm not sure I understand. The dataset is 3000 samples wide (`xn`) and the figure is 10 inches wide (`width`), giving an effective resolution of 300 dpi (if corresponding 1-to-1.) With dpi=72, the data needs to be resampled at a ratio of 25:6 = ~4.166667, which seems odd.

Comment (QuLogic) [inline examples/showcase/mandelbrot.py]: Oops, just realized I missed the /2, so it's really 1500 samples wide with effective resolution of 150 dpi, and resampling ratio of 25:12 = ~2.083333. But in that case, how about making it exactly 2 by setting dpi=75?

**Comment by tacaswell** [comment]
backported to v2.x as 1209e70 

Thanks @rougier

**Comment by jniediek** [comment]
I think there are several bugs in this example, which make it impossible to run it with python2.7, which breaks the documentation build process.

As I'm unfamiliar with matplotlib's policy for examples, I'll just list here what I consider to be bugs (and fixes):

    python2.7 mandelbrot.py
    File "mandelbrot.py", line 71
    SyntaxError: Non-ASCII character '\xe2' in file mandelbrot.py on line 71, but no encoding declared;     see http://python.org/dev/peps/pep-0263/ for details

Adding `# -*- coding: utf-8 -*-` as first line:

    python2.7 mandelbrot.py 
    Traceback (most recent call last):
    File "mandelbrot.py", line 70, in <module>
    major, minor, micro = matplotlib.__version__.split('.', maxsplit=2)
    TypeError: split() takes no keyword arguments

Changing to
    
    major, minor, micro = matplotlib.__version__.split('.')[:3]

And changing the next line to

    text = ("The Mandelbrot fractal set\n"
            "Rendered with matplotlib %s.%s, %s — http://matplotlib.org".decode('utf-8')
            % (major, minor, year))

It finally works with python2.7.

I have a feeling that using `six` and `from __future__ import` would be appropriate here, but as I said, I'm unfamiliar with the policies.

**Comment by tacaswell** [comment]
@jniediek already fixed: https://github.com/matplotlib/matplotlib/pull/7473

I strongly encourage you to move to python3 as soon as possible.

See https://python-3-for-scientists.readthedocs.io/en/latest/ and https://python3statement.github.io/ 

TLDR: You get nice stuff for switching and you are going start losing access to new versions of libraries in 2018-2020 time frame.
